feat: add `Client.query_and_wait` which directly returns a `RowIterator` of results

tswast opened the pull request:
Set the `QUERY_PREVIEW_ENABLED=TRUE` environment variable to use this with the new JOB_CREATION_OPTIONAL mode (currently in preview).

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Note: Currently in DRAFT mode as I realized that some of the necessary work to speed things up overlaps with the `query(api_method="QUERY")` work (https://github.com/googleapis/python-bigquery/pull/1723), so I'll work on modifying that first before moving on to the next step of a "jobless" API.

Towards #589 🦕

tswast committed:
feat: add `Client.query_and_wait` which directly returns a `RowIterator` of results

Set the `QUERY_PREVIEW_ENABLED=TRUE` environment variable to use this with the
new JOB_CREATION_OPTIONAL mode (currently in preview).

tswast committed:
perf: use the first page a results when `query(api_method="QUERY")`

tswast committed:
add tests

tswast committed:
respect max_results with cached page

tswast committed:
respect page_size, also avoid bqstorage if almost fully downloaded

tswast committed:
skip true test if bqstorage not installed

chalmerlowe commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -257,3 +287,48 @@ def do_query():
+    timeout: Optional[float],
+    job_retry: retries.Retry,
+) -> table.RowIterator:
+    """Initiate a query using jobs.query with jobCreationMode=JOB_CREATION_REQUIRED.
```
Should this be JOB_CREATION_OPTIONAL?
It seems that way based on the descriptions above.

tswast replied:
You're correct. I'll update.

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-RowIterator._is_completely_cached

tswast committed:
coverage

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-RowIterator._is_completely_cached

tswast committed:
implement basic query_and_wait and add code sample to test

tswast committed:
avoid duplicated QueryJob construction

tswast committed:
update unit tests

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

tswast committed:
fix merge conflict in rowiterator

tswast committed:
support max_results, add tests

tswast committed:
retry tests

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

tswast committed:
unit test coverage

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

tswast committed:
dont retry twice

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

chalmerlowe commented on google/cloud/bigquery/client.py:
```diff
@@ -3405,6 +3405,91 @@ def query(
+        page_size: Optional[int] = None,
+        max_results: Optional[int] = None,
+    ) -> RowIterator:
+        """Run the query, wait for it to finish, and return the results.
```
This docstring is more comprehensive than the docstring in `__job_helpers.py`. Is that intentional OR will the two docstrings be revised to provide a comprehensive description of the `query_and_wait()` function in both locations.

tswast replied:
Good point. Fixed! They now both have the same content. There was some important content not in the public version, so this makes sure that's now documented.

chalmerlowe reviewed (commented):
LGTM. Minor comment about docstrings.

tswast committed:
fix mypy_samples session

tswast committed:
consolidate docstrings for query_and_wait

tswast committed:
remove mention of job ID

tswast commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -123,7 +147,13 @@ def do_query():
 
 
-def _to_query_request(job_config: Optional[job.QueryJobConfig]) -> Dict[str, Any]:
+def _to_query_request(
```
If destination table is specified, we can't do `jobs.query`, need to fallback `jobs.insert`.

tswast commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -150,6 +180,15 @@ def _to_query_request(job_config: Optional[job.QueryJobConfig]) -> Dict[str, Any
 
+    if timeout is not None:
+        # Subtract a buffer for context switching, network latency, etc.
+        request_body["timeoutMs"] = max(0, int(1000 * timeout) - _TIMEOUT_BUFFER_MILLIS)
```
Note: this won't cancel the job. Need to provide a way for the user to cancel this job.

tswast commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -253,3 +292,158 @@ def do_query():
+    request_body = _to_query_request(
+        query=query, job_config=job_config, location=location, timeout=timeout
+    )
+
```
This is where fallback to jobs.insert needs to happen if unsupported parameters are passed in.

tswast commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -253,3 +292,158 @@ def do_query():
+    elif page_size is not None or max_results is not None:
+        request_body["maxResults"] = page_size or max_results
+
+    if os.getenv("QUERY_PREVIEW_ENABLED", "").casefold() == "true":
```
See: Java client changes. Needs to check a set of projects instead of all projects.

tswast replied:
@shollyman @mmladenovski Could you shared details about the requested `QUERY_PREVIEW_ENABLED` changes? I don't see any project ID comparisons here: https://github.com/googleapis/java-bigquery/blob/79959e5bbd84055575a2fde4e8ec8ed973a373e9/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryOptions.java#L136

mmladenovski replied:
@tswast , in our previous discussion, I only suggested for this environment variable to use a list of project IDs for a more fine-grained control. We have discussed it with @shollyman , and for the Java client we decided to go with a boolean. I am fine either way, staying consistent with a boolean or doing a more fine-grained project id list (if required by python client usage).

tswast replied:
I'll stick with the boolean for consistency with other client libraries, but definitely worth considering finer grained control in a future release if needed.

tswast commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -253,3 +292,158 @@ def do_query():
+
+        # For easier testing, handle the retries ourselves.
+        if retry is not None:
+            response = retry(client._call_api)(
```
Need to be careful with timeout here. `requestId` should help, but need to make sure the retry includes the requestId.

tswast committed:
fallback to jobs.insert for unsupported features

tswast committed:
distinguish API timeout from wait timeout

tswast committed:
add test for jobs.insert fallback

tswast committed:
populate default job config

tswast committed:
refactor default config

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

tswast committed:
add coverage for job_config

tswast committed:
cancel job if hasn't finished

tswast committed:
mypy

tswast committed:
allow unrealeased features in samples

tswast committed:
Merge branch 'main' into issue589-query_and_wait

tswast committed:
fix for 3.12

tswast committed:
Merge branch 'main' into issue589-query_and_wait

tswast committed:
Merge remote-tracking branch 'origin/main' into issue589-query_and_wait

tswast committed:
fix: keep `RowIterator.total_rows` populated after iteration

This was being reset in some cases when
the rows were all available in the
first page of results.

chalmerlowe commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -59,6 +83,25 @@ def make_job_id(job_id: Optional[str] = None, prefix: Optional[str] = None) -> s
+    if default_job_config is None:
+        return job_config
+
+    # anything that's not defined on the incoming
```
This comment is a bit unclear. Can we clean this up a bit?

chalmerlowe commented on google/cloud/bigquery/_job_helpers.py:
```diff
@@ -59,6 +83,25 @@ def make_job_id(job_id: Optional[str] = None, prefix: Optional[str] = None) -> s
+    # should be filled in with the default
+    # the incoming therefore has precedence
+    #
+    # Note that _fill_from_default doesn't mutate the receiver
```
For my own edification, what do we mean by "receiver"?

tswast committed:
Update google/cloud/bigquery/table.py

Co-authored-by: Anthonios Partheniou <partheniou@google.com>

chalmerlowe reviewed (approved):
LGTM. Please address minor comments.

tswast committed:
Merge remote-tracking branch 'origin/issue589-query_and_wait' into issue589-query_and_wait

tswast committed:
Merge remote-tracking branch 'origin/issue589-total_rows' into issue589-query_and_wait

tswast committed:
fix comments

tswast committed:
Merge branch 'main' into issue589-query_and_wait
